### DISCUSSION TITLE
Conditional variable not working

### DISCUSSION THREAD

**Discussion by worldpeace-germany** [missionpinball/missionpinball#83] [Q&A]
Hello,

I have a variable player like this

```
variable_player:
  s_pop_bumper_1_active:
    score: 100
  s_target_red_1_active:
    score: 1000
    potato: 1
  s_target_red_1_active{current_player.status_target_light_red_0 == 0}: 
      status_target_light_red_0:
        int: 1       
        action: set     
  s_target_red_1_active{current_player.status_target_light_red_0 == 1}: 
      status_target_light_red_0:
        int: 0
        action: set 
```

when my switch s_target_red_1 is being activated I want to toggle a variable called status_target_light_red_0, that variable should be either 0 or 1.

The problem now is that when the switch is activated that always both conditional entries are being executed, so the last in the list wins. In the above example the value is always 0, if I swap the order the value is always 1.

The log shows

```
INFO : SwitchController : <<<<<<< 's_target_red_1' inactive >>>>>>>
INFO : SwitchController : <<<<<<< 's_target_red_1' active >>>>>>>
INFO : EventManager : Event: ======'s_target_red_1_active'====== Args={}
INFO : EventManager : Event: ======'sw_playfield_active'====== Args={}
INFO : EventManager : Event: ======'machine_var_audits_switches_s_target_red_1'====== Args={'value': 219, 'prev_value': 218, 'change': 1}
INFO : EventManager : Event: ======'player_score'====== Args={'value': 5000, 'prev_value': 4000, 'change': 1000, 'player_num': 1, 'source': 'base'}
INFO : EventManager : Event: ======'player_potato'====== Args={'value': 5, 'prev_value': 4, 'change': 1, 'player_num': 1, 'source': 'base'}
INFO : EventManager : Event: ======'player_status_target_light_red_0'====== Args={'value': 1, 'prev_value': 0, 'change': 1, 'player_num': 1, 'source': 'base'}
INFO : EventManager : Event: ======'player_status_target_light_red_0'====== Args={'value': 0, 'prev_value': 1, 'change': -1, 'player_num': 1, 'source': 'base'}
EventManager: Event: ======'machine_var_audits_switches_s_target_red_1'====== Args={'value': 219, 'prev_value': 218, 'change': 1}
EventManager: Event: ======'player_score'====== Args={'value': 5000, 'prev_value': 4000, 'change': 1000, 'player_num': 1}
EventManager: Event: ======'player_potato'====== Args={'value': 5, 'prev_value': 4, 'change': 1, 'player_num': 1}
EventManager: Event: ======'player_status_target_light_red_0'====== Args={'value': 1, 'prev_value': 0, 'change': 1, 'player_num': 1}
EventManager: Event: ======'player_status_target_light_red_0'====== Args={'value': 0, 'prev_value': 1, 'change': -1, 'player_num': 1}
INFO : SwitchController : <<<<<<< 's_target_red_1' inactive >>>>>>>
```
As you can see both entries in the variable player are execute for each activation. What is wrong here?

Thanks!

**Comment by atummons** [comment]
If you look at your log, the switch is active at the start, and isn't inactive until your last line. This the argument is true. It is active, and the value is 0, so it makes it a 1. It is STILL active and is 1, so it makes it a 0.

I would use an event, not a switch state. An event happens 1 time, whereas a switch stays in a state for a period of time. Do you have a shot defined? If not, I would define one. Then trigger your logic off of target_red_1_hit or target_red_1_lit_hit or one of those varieties. You can find what is right for your desired outcome. But that should eliminate this issue for you.

**Reply by worldpeace-germany** [reply]
Hello, 

first of all thanks. I do agree that the switch activates in line 2 of the log and inactivates on the last line.

That is my switch config

```
switches:
  s_target_red_1:
    number: 0-0-27
    tags: playfield_active
```
thus

`s_target_red_1_active` is an event, this one [here](https://docs.missionpinball.org/en/latest/events/switch_active.html) . So I am using an event. The switch state I would consider to be this [device.switches.state](https://docs.missionpinball.org/en/latest/mechs/switches/index.html?highlight=switch%20state#monitorable-properties). What are you referring to?

I have now changed my config to us not the event from the switch itself, but my own event:

```
switches:
  s_target_red_1:
    number: 0-0-27
    events_when_activated: toggle_light
    tags: playfield_active
```

The log looks now like this

```
INFO : SwitchController : <<<<<<< 's_target_red_1' active >>>>>>>
INFO : EventManager : Event: ======'s_target_red_1_active'====== Args={}
INFO : EventManager : Event: ======'sw_playfield_active'====== Args={}
INFO : EventManager : Event: ======'toggle_light'====== Args={}
INFO : EventManager : Event: ======'machine_var_audits_switches_s_target_red_1'====== Args={'value': 233, 'prev_value': 232, 'change': 1}
INFO : EventManager : Event: ======'player_score'====== Args={'value': 2000, 'prev_value': 1000, 'change': 1000, 'player_num': 1, 'source': 'base'}
INFO : EventManager : Event: ======'player_potato'====== Args={'value': 2, 'prev_value': 1, 'change': 1, 'player_num': 1, 'source': 'base'}
INFO : EventManager : Event: ======'player_status_target_light_red_0'====== Args={'value': 1, 'prev_value': 0, 'change': 1, 'player_num': 1, 'source': 'base'}
INFO : EventManager : Event: ======'player_status_target_light_red_0'====== Args={'value': 0, 'prev_value': 1, 'change': -1, 'player_num': 1, 'source': 'base'}
EventManager: Event: ======'machine_var_audits_switches_s_target_red_1'====== Args={'value': 233, 'prev_value': 232, 'change': 1}
EventManager: Event: ======'player_score'====== Args={'value': 2000, 'prev_value': 1000, 'change': 1000, 'player_num': 1}
EventManager: Event: ======'player_potato'====== Args={'value': 2, 'prev_value': 1, 'change': 1, 'player_num': 1}
EventManager: Event: ======'player_status_target_light_red_0'====== Args={'value': 1, 'prev_value': 0, 'change': 1, 'player_num': 1}
EventManager: Event: ======'player_status_target_light_red_0'====== Args={'value': 0, 'prev_value': 1, 'change': -1, 'player_num': 1}
INFO : SwitchController : <<<<<<< 's_target_red_1' inactive >>>>>>>
```

the event toggle_light is being fired once. And again my variable is changed twice, from above log

```
INFO : EventManager : Event: ======'player_status_target_light_red_0'====== Args={'value': 1, 'prev_value': 0, 'change': 1, 'player_num': 1, 'source': 'base'}
INFO : EventManager : Event: ======'player_status_target_light_red_0'====== Args={'value': 0, 'prev_value': 1, 'change': -1, 'player_num': 1, 'source': 'base'}
```
Can you elaborate a bit on what I do have to do to get this working? I believe I did what you suggested, or I am not getting the point.

About shots, I did not have a shot defined yet. I did now

```
shots:
  my_first_shot:
    switch: s_target_red_1
```

```
variable_player:
(... some other entries here...)
  my_first_shot_hit{current_player.status_target_light_red_0 == 1}: 
      status_target_light_red_0:
        int: 0       
        action: set
  my_first_shot_hit{current_player.status_target_light_red_0 == 0}: 
      status_target_light_red_0:
        int: 1
        action: set  
```

Result: exactly the same, I do see the (shot)_hit event and again it is being executed twice.

Any idea is highly appreciated.

**Comment by worldpeace-germany** [comment]
Hi,

I have made a purely virtual example, maybe someone can run it and gets an idea what is wrong.

[Download zip here](https://easyupload.io/292rqp)

- run mpf and mpf monitor
- press s to start the game
- view variables in mpf monitor variable screen status_target_light_red_1, should show 0 right after start
- press z, view variable again, it shows 1
- press z, view variable again, it still shows 1 but should show zero

Thanks.

**Reply by Pucksnok** [reply]
It seems to work, it do change it to 0 and back to 1 instantly, if you change your order of the variable player "objects" it then it seems like it stays at 0 all the time but the truth is it go to 1 and instantly back to 0. If you only look at the variable it seems to not move at all, you have to check the logs.

INFO : SwitchController : <<<<<<< 's_target_red_1' active >>>>>>>
INFO : EventManager : Event: ======'toggle_light'====== Args={}
INFO : EventManager : Event: ======'machine_var_audits_switches_s_target_red_1'====== Args={'value': 190, 'prev_value': 189, 'change': 1}
INFO : EventManager : Event: ======'player_status_target_light_red_0'====== Args={'value': 0, 'prev_value': 1, 'change': -1, 'player_num': 1, 'source': 'base'}
INFO : EventManager : Event: ======'player_status_target_light_red_0'====== Args={'value': 1, 'prev_value': 0, 'change': 1, 'player_num': 1, 'source': 'base'}
EventManager: Event: ======'machine_var_audits_switches_s_target_red_1'====== Args={'value': 190, 'prev_value': 189, 'change': 1}
EventManager: Event: ======'player_status_target_light_red_0'====== Args={'value': 0, 'prev_value': 1, 'change': -1, 'player_num': 1}
EventManager: Event: ======'player_status_target_light_red_0'====== Args={'value': 1, 'prev_value': 0, 'change': 1, 'player_num': 1}
INFO : SwitchController : <<<<<<< 's_target_red_1' inactive >>>>>>>

```
  toggle_light{current_player.status_target_light_red_0==1}: 
      status_target_light_red_0:
        int: 0       
        action: set  
  toggle_light{current_player.status_target_light_red_0==0}: 
      status_target_light_red_0:
        int: 1       
        action: set 
```


I'm not sure why this is happening since I'm new to this but this solution below works without any issues.

```
event_player:
  toggle_light{current_player.status_target_light_red_0==1}: status_target_set_0
  toggle_light{current_player.status_target_light_red_0==0}: status_target_set_1

variable_player:
  status_target_set_0:
      status_target_light_red_0:
        int: 0       
        action: set  
  status_target_set_1:
      status_target_light_red_0:
        int: 1       
        action: set 
```

**Reply by worldpeace-germany** [reply]
Thanks, your proposal is working fine for me. In other words basically solved. But

It seems that the conditions work for the event_player, I saw examples in the documentation for slide_player and light_player.

So either it is not supposed to work for a variable_player, then this should be mentioned clearly in the documentation. Or it is suppose to work, then it is a bug which should get fixed. 

Does someone know if it is "works as designed" or a bug? In the first case I would update the documentation, in the second case I would ask if someone know what part of the source code is handling that, then I would be willing to see if I can do something to fix it.

**Reply by Pucksnok** [reply]
Conditions work for variable player and your code did change the variable as well, the problem seemed to be it did the two conditions on the same event as it were active and they both fulfilled the condition set... So they took each other out or how to explain it, it changed the variable and immediately changed it back when the second condition were met. Checking the logs what actually happen is always a good start. If you believe it's a bug, start a issue.

**Reply by worldpeace-germany** [reply]
I fully agree what you write about the log, see my original post

> 
> The problem now is that when the switch is activated that always both conditional entries are being executed, so the last in the list wins. In the above example the value is always 0, if I swap the order the value is always 1.
> 
> The log shows
> 
> [log removed in citation]
> 
> As you can see both entries in the variable player are execute for each activation. 

Since swapping the order shows as result that always the last entry wins, it looks to me mpf goes through the variable player and executes all entries for the given event without obeying the given condition.

I will open an issue report.

https://github.com/missionpinball/mpf/issues/1685

**Comment by avanwinkle** [comment]
Closing this discussion, see the linked issue for more details.

In short, the variable player executes all the entries sequentially and the first one changes the player variable to the value that then makes the second one true. So this is not an issue with the variable player, but the configuration of having conditional variable players change the value of the condition while they are simultaneously being evaluated.
